### PR TITLE
Corda-1869 serialisation refactor

### PR DESCRIPTION
**Pull request by distributedleetravis**
A few small sets of changes related to https://r3-cev.atlassian.net/browse/CORDA-1869, mainly the result of tidying up code as I was reading it in order to understand it better.

The first set of changes is a fairly trivial refactoring of the `inferTypeVariables` method, aimed at improving readability.

The second set of changes is the result of changing the extension method `Type.asClass()` so that its return type is `Class<*>` rather than `Class<*>?`, because it can never return `null` and was only suggesting that it might because Kotlin can't ensure that the list of possible cases given to its `when` clause is exhaustive. Instead of returning `null` on the `else` case, we offer to throw an `UnsupportedOperationException`. Having done this, it was then possible to eliminate a lot of unnecessary null checks and not-null assertions.

The third set of changes is a refactor of the `FingerPrinter` interface and `SerializerFingerPrinter` implementation. Three changes are involved:

1) We remove the `setFactory` method from `FingerPrinter`, and insist that `SerializerFingerPrinter` is initialised with the factory it is going to need. We can now also insist that this factory is never `null`, without resorting to `lateinit`.
2) We break up some deeply-nested logic to make the execution flow a bit easier to follow.
3) We pull out the working state of the fingerprinter into a set of fields in their own class, `FingerPrinterState`, instead of passing it down in pieces through a chain of method calls. Because the recursion never needs to backtrack, we don't need to be able to pull previous state values out of the stack - it's safe just to mutate the current state variables freely as we go along. The `debugIndent` parameter, which is never used anywhere, disappears altogether.

Finally, there is some general tidying and refactoring of code around `PropertyDescriptor`s, including a minor bugfix (we should be checking the first parameter type, not the return type, of setter methods when deciding which one is the least generic).

### CODE REVIEW

File: serialization/src/main/kotlin/net/corda/serialization/internal/amqp/SerializationHelper.kt [hunk 1]
```diff
@@ -75,145 +67,13 @@ fun constructorForDeserialization(type: Type): KFunction<Any>? {
...
-                        // take into account those constructor properties that don't directly map to a named
-                        // property which are, by default, already added to the map
-                        classProperties.computeIfAbsent(groups[2]!!.value) { PropertyDescriptor() }
-
-                properties.apply {
-                    when (groups[1]!!.value) {
-                        "set" -> {
-                            if (func.parameterCount == 1) {
-                                if (setter == null) setter = func
-                                else if (TypeToken.of(setter!!.genericReturnType).isSupertypeOf(func.genericReturnType)) {
```
Comment (distributedleetravis) [inline serialization/src/main/kotlin/net/corda/serialization/internal/amqp/SerializationHelper.kt]: Tiny bug: we ought to be comparing the first generic parameter type, not the return type.

File: serialization/src/main/kotlin/net/corda/serialization/internal/amqp/PropertyDescriptor.kt [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+// Map a sequence of fields by field name.
+private fun Sequence<Field>.byFieldName() = map { it.name to it }.toMap()
+
+// Select only those methods that are public (and are not the "getClass" method)
+private fun Sequence<Method>.thatArePublic() = filter { it.isPublic && it.name != "getClass" }
+
+// Select only those methods that are isX/getX/setX methods
+private fun Sequence<Method>.thatArePropertyMethods() = map { method ->
+    propertyMethodRegex.find(method.name)?.let {
```
Comment (fenryka) [inline serialization/src/main/kotlin/net/corda/serialization/internal/amqp/PropertyDescriptor.kt]: name implicit param

File: serialization/src/main/kotlin/net/corda/serialization/internal/amqp/SerializationHelper.kt (hunk 1 above)
Comment (fenryka) [reply serialization/src/main/kotlin/net/corda/serialization/internal/amqp/SerializationHelper.kt]: good catch

File: serialization/src/main/kotlin/net/corda/serialization/internal/amqp/SerializationHelper.kt [hunk 3]
```diff
@@ -235,174 +95,132 @@ internal fun <T : Any> propertiesForSerializationFromConstructor(
...
+
+    // We will already have disambiguated getA for property A or a but we still need to cope
+    // with the case we don't know the case of A when the parameter doesn't match a property
+    // but has a getter
+    val matchingProperty = classProperties[name] ?: classProperties[name.capitalize()]
+    ?: throw AMQPNotSerializableException(type,
+            "Constructor parameter - \"$name\" -  doesn't refer to a property of \"$clazz\"")
+
+    // If the property has a getter we'll use that to retrieve it's value from the instance, if it doesn't
+    // *for *know* we switch to a reflection based method
```
Comment (fenryka) [inline serialization/src/main/kotlin/net/corda/serialization/internal/amqp/SerializationHelper.kt]: *now* not *know* (which to be fair is probably my typo but... you're moving the code :)

**Comment by mikehearn** [comment]
This PR should probably be re-reviewed now you have added more commits, if they aren't addressing review comments. Better to get this PR merged and then start a new one, I think.

**Comment by fenryka** [comment]
Remember to squash :)

**Pull request merged**